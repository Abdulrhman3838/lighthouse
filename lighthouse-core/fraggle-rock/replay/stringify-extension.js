/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

const {isNavigationStep} = require('./replay-helpers.js');
const {PuppeteerStringifyExtension} = require('@puppeteer/replay');

class LighthouseStringifyExtension extends PuppeteerStringifyExtension {
  isTimespanRunning = false;

  /**
   * @param {Parameters<PuppeteerStringifyExtension['beforeAllSteps']>} args
   */
  async beforeAllSteps(...args) {
    const [out, flow] = args;
    out.appendLine(`const fs = require('fs');`);
    out.appendLine(`const lhApi = require('lighthouse/lighthouse-core/fraggle-rock/api.js');`);

    let isMobile = true;
    for (const step of flow.steps) {
      if (step.type !== 'setViewport') continue;
      isMobile = step.isMobile;
    }

    if (isMobile) {
      // eslint-disable-next-line max-len
      out.appendLine(`const config = undefined;`);
    } else {
      // eslint-disable-next-line max-len
      out.appendLine(`const config = require('lighthouse/lighthouse-core/config/desktop-config.js');`);
    }

    await super.beforeAllSteps(...args);

    const configContext = {
      settingsOverrides: {
        screenEmulation: {
          disabled: true,
        },
      },
    };

    // eslint-disable-next-line max-len
    out.appendLine(`const lhFlow = await lhApi.startFlow(page, {name: '${flow.title || 'undefined'}', config, configContext: ${JSON.stringify(configContext)}});`);
  }

  /**
   * @param {Parameters<PuppeteerStringifyExtension['stringifyStep']>} args
   */
  async stringifyStep(...args) {
    const [out, step] = args;

    if (step.type === 'setViewport') {
      await super.stringifyStep(...args);
      return;
    }

    const isNavigation = isNavigationStep(step);

    if (isNavigation) {
      if (this.isTimespanRunning) {
        out.appendLine(`await lhFlow.endTimespan();`);
        this.isTimespanRunning = false;
      }
      out.appendLine(`await lhFlow.startNavigation();`);
    } else if (!this.isTimespanRunning) {
      out.appendLine(`await lhFlow.startTimespan();`);
      this.isTimespanRunning = true;
    }

    await super.stringifyStep(...args);

    if (isNavigation) {
      out.appendLine(`await lhFlow.endNavigation();`);
    }
  }

  /**
   * @param {Parameters<PuppeteerStringifyExtension['afterAllSteps']>} args
   */
  async afterAllSteps(...args) {
    const [out] = args;
    if (this.isTimespanRunning) {
      out.appendLine(`await lhFlow.endTimespan();`);
    }
    out.appendLine(`const lhFlowReport = await lhFlow.generateReport();`);
    out.appendLine(`fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)`);
    await super.afterAllSteps(...args);
  }
}

module.exports = LighthouseStringifyExtension;
