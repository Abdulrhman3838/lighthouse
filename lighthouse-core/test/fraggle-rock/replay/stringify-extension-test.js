/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

import {promises as fs} from 'fs';
import {promisify} from 'util';
import {execFile} from 'child_process';

import {stringify} from '@puppeteer/replay';

import {LH_ROOT} from '../../../../root.js';
import LighthouseStringifyExtension from '../../../fraggle-rock/replay/stringify-extension.js';
import {getAuditsBreakdown, createTestState} from '../scenarios/pptr-test-utils.js';
import {readJson} from '../../test-utils.js';

const execFileAsync = promisify(execFile);
const replayFlowJson = readJson(`${LH_ROOT}/lighthouse-core/test/fixtures/fraggle-rock/replay/test-flow.json`);
const FLOW_JSON_REGEX = /window\.__LIGHTHOUSE_FLOW_JSON__ = (.*);<\/script>/;

describe('LighthouseStringifyExtension', function() {
  // eslint-disable-next-line no-invalid-this
  this.timeout(90_000);

  const state = createTestState();
  state.installServerHooks();

  const tmpDir = `${LH_ROOT}/.tmp/replay`;
  let testTmpDir = '';
  let scriptPath = '';

  before(async () => {
    await fs.mkdir(tmpDir, {recursive: true});
    // Stringified exports are CJS
    fs.writeFile(`${tmpDir}/package.json`, '{"type": "commonjs"}');
  });

  beforeEach(async () => {
    testTmpDir = await fs.mkdtemp(`${tmpDir}/replay-`);
    scriptPath = `${testTmpDir}/stringified.js`;
  });

  after(async () => {
    await fs.rm(tmpDir, {recursive: true, force: true});
  });

  it('crates a valid desktop script', async () => {
    const scriptContents = await stringify(replayFlowJson, {
      extension: new LighthouseStringifyExtension(),
    });

    expect(scriptContents).toMatchSnapshot();
    await fs.writeFile(scriptPath, scriptContents);

    await execFileAsync('node', [scriptPath]);

    const reportHtml = await fs.readFile(`${testTmpDir}/flow.report.html`, 'utf-8');
    const flowResultJson = FLOW_JSON_REGEX.exec(reportHtml)?.[1];
    if (!flowResultJson) throw new Error('Could not find flow json');

    /** @type {LH.FlowResult} */
    const flowResult = JSON.parse(flowResultJson);
    expect(flowResult.steps).toHaveLength(4);
    expect(flowResult.name).toEqual(replayFlowJson.title);

    for (const {lhr} of flowResult.steps) {
      expect(lhr.configSettings.formFactor).toEqual('desktop');
      expect(lhr.configSettings.screenEmulation.disabled).toBeTruthy();

      const {auditResults, erroredAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.length).toBeGreaterThanOrEqual(10);
      // TODO: INP breakdown diagnostic audit is broken because of old Chrome
      expect(erroredAudits.length).toBeLessThanOrEqual(1);
    }
  });
});
