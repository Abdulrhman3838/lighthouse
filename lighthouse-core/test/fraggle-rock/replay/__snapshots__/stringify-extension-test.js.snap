// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LighthouseStringifyExtension crates a valid desktop script 1`] = `
"const fs = require('fs');
const lhApi = require('lighthouse/lighthouse-core/fraggle-rock/api.js');
const config = require('lighthouse/lighthouse-core/config/desktop-config.js');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const lhFlow = await lhApi.startFlow(page, {name: 'web.dev', config, configContext: {\\"settingsOverrides\\":{\\"screenEmulation\\":{\\"disabled\\":true}}}});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":990,\\"height\\":900})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"https://web.dev/\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"#sub-newsletter\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 9,
        y: 9.671875,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"#sub-pii-spii\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 8,
        y: 13.3203125,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/First Name\\"],[\\"#sub-firstname\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 182.859375,
        y: 36.25,
      },
    });
  }
  await lhFlow.endTimespan();
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    const element = await waitForSelectors([[\\"#main > div > div:nth-child(5) > div > div.headline.all-center.flow > div > a\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 25.703125,
        y: 15.3515625,
      },
    });
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await browser.close();

  async function waitForSelectors(selectors, frame, options) {
    for (const selector of selectors) {
      try {
        return await waitForSelector(selector, frame, options);
      } catch (err) {
        console.error(err);
      }
    }
    throw new Error('Could not find element for selectors: ' + JSON.stringify(selectors));
  }

  async function scrollIntoViewIfNeeded(element, timeout) {
    await waitForConnected(element, timeout);
    const isInViewport = await element.isIntersectingViewport({threshold: 0});
    if (isInViewport) {
      return;
    }
    await element.evaluate(element => {
      element.scrollIntoView({
        block: 'center',
        inline: 'center',
        behavior: 'auto',
      });
    });
    await waitForInViewport(element, timeout);
  }

  async function waitForConnected(element, timeout) {
    await waitForFunction(async () => {
      return await element.getProperty('isConnected');
    }, timeout);
  }

  async function waitForInViewport(element, timeout) {
    await waitForFunction(async () => {
      return await element.isIntersectingViewport({threshold: 0});
    }, timeout);
  }

  async function waitForSelector(selector, frame, options) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to waitForSelector');
    }
    let element = null;
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (element) {
        element = await element.waitForSelector(part, options);
      } else {
        element = await frame.waitForSelector(part, options);
      }
      if (!element) {
        throw new Error('Could not find element: ' + selector.join('>>'));
      }
      if (i < selector.length - 1) {
        element = (await element.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
      }
    }
    if (!element) {
      throw new Error('Could not find element: ' + selector.join('|'));
    }
    return element;
  }

  async function waitForElement(step, frame, timeout) {
    const count = step.count || 1;
    const operator = step.operator || '>=';
    const comp = {
      '==': (a, b) => a === b,
      '>=': (a, b) => a >= b,
      '<=': (a, b) => a <= b,
    };
    const compFn = comp[operator];
    await waitForFunction(async () => {
      const elements = await querySelectorsAll(step.selectors, frame);
      return compFn(elements.length, count);
    }, timeout);
  }

  async function querySelectorsAll(selectors, frame) {
    for (const selector of selectors) {
      const result = await querySelectorAll(selector, frame);
      if (result.length) {
        return result;
      }
    }
    return [];
  }

  async function querySelectorAll(selector, frame) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to querySelectorAll');
    }
    let elements = [];
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (i === 0) {
        elements = await frame.$$(part);
      } else {
        const tmpElements = elements;
        elements = [];
        for (const el of tmpElements) {
          elements.push(...(await el.$$(part)));
        }
      }
      if (elements.length === 0) {
        return [];
      }
      if (i < selector.length - 1) {
        const tmpElements = [];
        for (const el of elements) {
          const newEl = (await el.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
          if (newEl) {
            tmpElements.push(newEl);
          }
        }
        elements = tmpElements;
      }
    }
    return elements;
  }

  async function waitForFunction(fn, timeout) {
    let isActive = true;
    setTimeout(() => {
      isActive = false;
    }, timeout);
    while (isActive) {
      const result = await fn();
      if (result) {
        return;
      }
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    throw new Error('Timed out');
  }
})();
"
`;
